### DISCUSSION TITLE
Differences in casting date in sqlite and mssql

### DISCUSSION THREAD

**Discussion by jahan01** [sqlalchemy/sqlalchemy#7614] [Usage Questions] [answered]
I am writing an application for mssql backend and unit-test runs on sqlite. At the moment our CI cannot host sql server, hence unit tests have to run with sqlite.

One of the selectable query has to cast a column to date and the dialect changes depending on the backend.

mssql selectable:
```python
select(
    cast(Table1.mydate, Date).label("Date"),
    Table1.id.label("identifier"),
    ...
).where(...)
```

sqlite version:
```python
select(
    func.DATE(Table1.mydate).label("Date"),
    Table1.id.label("identifier"),
    ...
).where(...)
```

Is there a way to conditionally construct this selectable? i.e., use `CAST(..` if mssql or `func.DATE` if sqlite ?

Sqlalchemy version: 1.4

**Comment by zzzeek** [comment, marked as answer]
the approach right now is to make your own `@compiles` construct for these situations.  it's not the most convenient approach but for now it works well, as this is a common issue with dates where all the DBs have completely different syntaxes.  see the example at https://docs.sqlalchemy.org/en/14/core/compiler.html#utc-timestamp-function for a relatively quick way to make a function that renders in a custom way for different backends.

**Reply by zzzeek** [reply]
if you use FunctionElement, the arguments are there under .clauses, see the next example https://docs.sqlalchemy.org/en/14/core/compiler.html#greatest-function for how to process these

**Reply by jahan01** [reply]
thanks, this helped a lot. I ended up with following code.

```python
class CastDate(expression.FunctionElement):
    type = DateTime()
    inherit_cache = True


@compiles(CastDate)
def cast_date_default(element, compiler, **kw):
    date = compiler.process(element.clauses, **kw)
    return f"CAST({date} AS DATE)"


@compiles(CastDate, "sqlite")
def cast_date_sqlite(element, compiler, **kw):
    date = compiler.process(element.clauses, **kw)
    return compiler.process(func.DATE(date))
```

I was originally thinking in terms of hybrid attributes with expression, but none of the [examples](https://docs.sqlalchemy.org/en/14/orm/extensions/hybrid.html) had per dialect expression. So dint know how to achieve that.

**Reply by CaselIT** [reply]
for sqlite you can probably use
```py
@compiles(CastDate, "sqlite")
def cast_date_sqlite(element, compiler, **kw):
    return compiler.process(func.DATE(element.clauses), **kw)
```

**Reply by jahan01** [reply]
@CaselIT : i just had tried your suggestion and it doesnt seem to work. I get following error:
```
sqlalchemy.exc.ArgumentError: SQL expression element expected, got <sqlalchemy.sql.elements.ClauseList object at 0x10ca08dc0>
```

Btw, I am using v1.4.27.

**Reply by jahan01** [reply]
somehow the below version produces a `None` result for sqlite:
```python
@compiles(CastDate, "sqlite")
def cast_date_sqlite(element, compiler, **kw):
    date = compiler.process(element.clauses, **kw)
    return compiler.process(func.DATE(date))
```

So I had to fix by surrounding `sqlalchemy.sql.expression.text`, but I dont understand the reason though.
Whoever is reading this later my previous comment doesn't fully work.
Below is the working code for sqlite.
```python
@compiles(CastDate, "sqlite")
def cast_date_sqlite(element, compiler, **kw):
    date = compiler.process(element.clauses, **kw)
    return compiler.process(func.DATE(text(date)))
```

**Reply by zzzeek** [reply]
well "date" becomes a string so if you are trying to re-use that in another SQL element, yes you'd have to use text.

though it should work more directly as:

```py
@compiles(CastDate, "sqlite")
def cast_date_sqlite(element, compiler, **kw):
    return compiler.process(func.DATE(*element.clauses), **kw)
```

**Reply by jahan01** [reply]
thanks for explanation.

And your simplified version for sqlite works too. Earlier I wasn't unpacking the `element.clauses`
